### PR TITLE
Refactor analytical temperature field for laser impact

### LINKED ISSUE #202: Move LaserOperation::compute_analytical_temperature_field to own class

### PR DESCRIPTION
**Pull request by mschreter**
closes #202 

@nmuch: FYI

### CODE REVIEW

File: include/meltpooldg/heat/laser_analytical_temperature_field.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    LaserAnalyticalTemperatureField(const LaserData<double>::AnalyticalData &data_in,
+                                    const MaterialData<double> &             material_in,
+                                    const double                             scan_speed);
+
+    void
+    compute_temperature_field(const ScratchData<dim> &scratch_data,
+                              const VectorType &      level_set_as_heaviside,
+                              VectorType &            temperature,
+                              const unsigned int      temp_dof_idx,
+                              const double            laser_power,
```
Comment (peterrum) [inline include/meltpooldg/heat/laser_analytical_temperature_field.hpp:30]: Why not pass these to the constructor (except the vectors)?

File: include/meltpooldg/heat/laser_analytical_temperature_field.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    using VectorType = LinearAlgebra::distributed::Vector<double>;
+    /**
+     * The temperature function below is derived from the publication on
+     * "Heat Source Modeling in Selective Laser Melting" by E. Mirkoohi, D. E. Seivers,
+     *  H. Garmestani and S. Y. Liang
+     */
+    double
+    local_compute_temperature_field(Point<dim>        point,
+                                    const double      heaviside,
+                                    const double      laser_power,
```
Comment (peterrum) [inline include/meltpooldg/heat/laser_analytical_temperature_field.hpp:43]: Same here.

File: include/meltpooldg/heat/laser_analytical_temperature_field.hpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  private:
+    using VectorType = LinearAlgebra::distributed::Vector<double>;
+    /**
+     * The temperature function below is derived from the publication on
+     * "Heat Source Modeling in Selective Laser Melting" by E. Mirkoohi, D. E. Seivers,
+     *  H. Garmestani and S. Y. Liang
+     */
+    double
+    local_compute_temperature_field(Point<dim>        point,
```
Comment (peterrum) [inline include/meltpooldg/heat/laser_analytical_temperature_field.hpp:41]:
```suggestion
    local_compute_temperature_field(cont Point<dim> &       point,
```

File: source/heat/laser_analytical_temperature_field.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    temperature.compress(VectorOperation::insert);
+    scratch_data.get_constraint(temp_dof_idx).distribute(temperature);
+    level_set_as_heaviside.zero_out_ghost_values();
+  }
+
+  template <int dim>
+  double
+  LaserAnalyticalTemperatureField<dim>::local_compute_temperature_field(
+    Point<dim>        point,
```
Comment (peterrum) [inline source/heat/laser_analytical_temperature_field.cpp:57]:
```suggestion
    const Point<dim> &       point,
```

File: source/heat/laser_analytical_temperature_field.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    const double      heaviside,
+    const double      laser_power,
+    const Point<dim> &laser_position) const
+  {
+    const double &P  = laser_power;
+    const double &v  = scan_speed;
+    const double &T0 = laser_data.ambient_temperature;
```
Comment (peterrum) [inline source/heat/laser_analytical_temperature_field.cpp:64]:
```suggestion
    const double P  = laser_power;
    const double v  = scan_speed;
    const double T0 = laser_data.ambient_temperature;
```

File: source/heat/laser_analytical_temperature_field.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    double weight;
+
+    if (laser_data.variable_properties_over_interface)
+      weight = heaviside;
+    else
+      weight = (heaviside > 0.5) ? 1.0 : 0.0;
```
Comment (peterrum) [inline source/heat/laser_analytical_temperature_field.cpp:71]: I like if variables are constant :)
```suggestion
    const double weight = laser_data.variable_properties_over_interface ? 
      heaviside :  ((heaviside > 0.5) ? 1.0 : 0.0);
```
... but not everyone likes the "?" operator :(

File: source/heat/laser_analytical_temperature_field.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    const double capacity =
+      UtilityFunctions::interpolate(weight, material.first.capacity, material.second.capacity);
+
+    const double density =
+      material.first.density + weight * (material.second.density - material.first.density);
+    const double thermal_diffusivity = conductivity / (density * capacity);
+
+    // modify temperature profile to be anisotropic
+    for (int d = 0; d < dim - 1; d++)
+      point[d] *= laser_data.temperature_x_to_y_ratio;
```
Comment (peterrum) [inline source/heat/laser_analytical_temperature_field.cpp:89]: If point is const, you would need to create a new point here.

**Review by peterrum — APPROVED** [review]
Some minor comments.

File: include/meltpooldg/heat/laser_analytical_temperature_field.hpp (hunk 2 above)
Comment (mschreter) [reply include/meltpooldg/heat/laser_analytical_temperature_field.hpp:43]: laser_power may change throughout the simulation, that's why I left it inside the function arguments.

**Review by peterrum — APPROVED** [review]
Thx!

**Comment by mschreter** [comment]
Thanks @peterrum for your review :-).

**Pull request merged**